### PR TITLE
Return Total Count with ListResources

### PR DESCRIPTION
**Pull request by kimlisa**
#### Description
The web UI needed to show the total count of resources available after filter (if any). In order to do this, we use the same path we take when sorting is requested, where we retrieve the entire list upfront.

I changed the `ListResources` response to a struct to make adding extra fields easier. This shouldn't affect backwards compatibility b/c I did not change the grpc message (just adding an additional field). A large part of my changes are b/c of this return value change.

I also had to edit how `ListResources` was working in the `auth_with_roles` layer which is explained by the comment I put in the code.

### CODE REVIEW

File: lib/services/local/presence.go [hunk 1]
```diff
@@ -1700,23 +1704,45 @@ func FakePaginate(resources []types.ResourceWithLabels, req proto.ListResourcesR
 		PredicateExpression: req.PredicateExpression,
 	}
 
+	// afterMatchCount counts matches after we found the limit.
+	afterMatchCount := 0
+	// filterAll is a flag to continue matching even after we found limit,
+	// to determine the total count. This flag is only enabled on the first fetch.
+	filterAll := req.NeedTotalCount && req.StartKey == ""
```
Comment (espadolini) [inline lib/services/local/presence.go:1711]: Why the check for `StartKey` if the caller has asked for the total count?

File: lib/services/local/presence.go [hunk 2]
```diff
@@ -1700,23 +1704,45 @@ func FakePaginate(resources []types.ResourceWithLabels, req proto.ListResourcesR
 		PredicateExpression: req.PredicateExpression,
 	}
 
+	// afterMatchCount counts matches after we found the limit.
+	afterMatchCount := 0
```
Comment (espadolini) [inline lib/services/local/presence.go:1708]: I think the logic would be simpler if this was just a `matchCount` that was zeroed when breaking early from the loop.

File: lib/auth/auth_with_roles.go [hunk 3]
```diff
@@ -973,6 +996,64 @@ func (a *ServerWithRoles) checkAccessToResource(resource types.Resource) error {
...
+		}
+
+		servers := types.DatabaseServers(dbservers)
+		if err := servers.SortByCustom(req.SortBy); err != nil {
+			return nil, trace.Wrap(err)
+		}
+		resources = servers.AsResources()
+
+	default:
+		return nil, trace.NotImplemented("resource type %q is not supported for getResources", req.ResourceType)
```
Comment (espadolini) [inline lib/auth/auth_with_roles.go:1045]:
```suggestion
		return nil, trace.NotImplemented("resource type %q is not supported for listResourcesWithSort", req.ResourceType)
```

File: api/client/proto/authservice.proto [hunk 4]
```diff
@@ -1497,6 +1500,8 @@ message ListResourcesResponse {
     // continue retrieving pages of resource. If NextKey is empty, there are no
     // more pages.
     string NextKey = 2 [ (gogoproto.jsontag) = "next_key,omitempty" ];
+    // TotalCount is the total number of resources available after filter, if any.
```
Comment (espadolini) [inline api/client/proto/authservice.proto:1503]: "if any" but also if the total count was asked for and the request was legitimate?

File: api/client/proto/authservice.proto [hunk 5]
```diff
@@ -1487,6 +1487,9 @@ message ListResourcesRequest {
     // SortBy describes which resource field and which direction to sort by.
     types.SortBy SortBy = 8
         [ (gogoproto.nullable) = false, (gogoproto.jsontag) = "sort_by,omitempty" ];
+    // NeedTotalCount is a flag when true, returns total number of resources available after filter.
```
Comment (espadolini) [inline api/client/proto/authservice.proto:1490]: style?
```suggestion
    // NeedTotalCount indicates whether or not the caller also wants the total number of resources after filtering.
```

File: api/types/resource.go [hunk 6]
```diff
@@ -360,3 +360,13 @@ func stringCompare(a string, b string, isDesc bool) bool {
 	}
 	return a < b
 }
+
+// ListResourcesResponse describes a non proto response to ListResouurces.
```
Comment (espadolini) [inline api/types/resource.go:364]:
```suggestion
// ListResourcesResponse describes a non proto response to ListResources.
```

File: lib/services/local/presence.go (hunk 2 above)
Comment (kimlisa) [reply lib/services/local/presence.go]: i had to zero it afterwards to handle the case where if loop ends first (but i did change the logic a bit to make it cleaner)

File: lib/services/local/presence.go (hunk 1 above)
Comment (kimlisa) [reply lib/services/local/presence.go]: you're right, the caller can toggle this flag. i changed to always return count if flag enabled

**Review by espadolini — APPROVED** [review]
The logic in `FakePaginate` is still a bit confusing but I tried rewriting it and I couldn't come up with anything simpler so maybe it's just inherently awkward to write this sort of code?

**Review by rosstimothy — APPROVED** [review]

**Review by zmb3 — APPROVED** [review]

**Review by rudream — APPROVED** [review]

**Pull request merged**